warehouse/query-core/src/main/java/datawave/query/jexl/functions/GroupingRequiredFilterFunctions.java

```diff
@@ -211,122 +242,138 @@
                 }
             }
         }
         
         // if there was a match found at all levels, then the matches.size will be equal to the
         // number of field/regex pairs
-        if (matches.size() < args.length / 2) { // truncated in case args.length was odd
-            matches.clear();
+        if (allMatches.size() < args.length / 2) { // truncated in case args.length was odd
+            allMatches.clear();
+        }
+        if (log.isTraceEnabled()) {
+            log.trace("returning matches:" + allMatches);
+        }
+        return Collections.unmodifiableCollection(allMatches);
+    }
+    
+    private static void manageMatchesInGroupLeftRemainingArgs(Object fieldValue, String regex, Collection<ValueTuple> allMatches, String theFirstMatch,
+                    String theNextMatch, ValueTuple currentMatch) {
+        
+        if (theNextMatch != null && theNextMatch.equals(theFirstMatch)) {
+            if (log.isTraceEnabled()) {
+                log.trace("\tfirst match equals the second: " + theFirstMatch + " == " + theNextMatch);
+            }
+            allMatches.addAll(EvaluationPhaseFilterFunctions.includeRegex(fieldValue, regex));
+            allMatches.add(currentMatch);
         }
-        log.trace("returning matches:" + matches);
-        return Collections.unmodifiableCollection(matches);
     }
     
-    public static Collection<ValueTuple> atomValuesMatch(Object... args) {
+    /**
+     * test that fields (field names) have values that match within the same grouping context.
+     * 
+     * @param fields
+     * @return a collection of matches
+     * */
+    public static Collection<ValueTuple> atomValuesMatch(Object... fields) {
         List<Iterable<?>> iterableFields = new ArrayList<>();
         
-        for (Object field : args) {
+        for (Object field : fields) {
             if (field instanceof Iterable == false) {
                 field = Collections.singleton(field);
             }
             iterableFields.add((Iterable<?>) field);
         }
-        return atomValuesMatch(iterableFields);
+        return atomValuesMatch(iterableFields.toArray(new Iterable[iterableFields.size()]));
     }
     
     /**
-     * test that fields have values that match within the same grouping context. QueryIterator will add (then remove later) the grouping context if it was not
-     * specified in the original query
+     * test that fields have values that match within the same grouping context.
      * 
      * @param fields
      * @return a collection of matches
      */
-    public static Collection<ValueTuple> atomValuesMatch(List<Iterable<?>> fields) {
-        Set<ValueTuple> matches = Sets.newHashSet();
-        if (fields.contains(null)) {
+    public static Collection<ValueTuple> atomValuesMatch(Iterable<?>... fields) {
+        Set<ValueTuple> matches = new HashSet<>();
+        if (fields.length == 0 || Arrays.asList(fields).contains(null)) {
```

Comment:
How often is this called? Would it be better to write a small function that the compiler would inline that iterates through fields looking for null instead of creating a list object (which you must still iterate through) that gets thrown away right away?